### DISCUSSION TITLE
Make 'end" of SimpleSpan nonzero

### DISCUSSION THREAD

**Discussion by purpl3F0x** [zesterer/chumsky#661] [Ideas]
Would it be more appropriate for `end` filed of the `SimpleSpan` be non-zero, so something like:
```rust
pub struct SimpleSpan<T = usize>
    pub start: T,
    pub end: std::num::NonZero<T>,
}
```

By definition, on the docs, the end is exclusive, so (0,0) values should not be valid. 
This will have the added benefit, of `Option<SimpleSpan>`, not requiring an extra byte. 

(https://play.rust-lang.org/?gist=63b58e61f659c6613cf1bba8035def00)

**Comment by zesterer** [comment]
`(0, 0)` is a valid span because spans can be (and frequently are) zero-length. Consider the following:

```rs
let function_call = ident
    .then(arg.repeated().collect().map(|args, e| (args, e.span())
        .delimited_by(lparen, rparen));

function_call.parse("foo()");
```

What span should the arguments receive? It would not be correct to have the span select the parentheses, since they surround the arguments but are themselves not the arguments. The correct span is a zero-width one that fits between the parentheses.

**Comment by purpl3F0x** [comment]
I though this could be something like `(4, 4)`, which is zero length also.
But your point is still valid, for other examples. Something similar "args" could be at the start of the string, let's say a C-like include for example, or a docstring
